diff --git a/apps/openmw/mwgui/windowmanagerimp.cpp b/apps/openmw/mwgui/windowmanagerimp.cpp
index 261a5e6e16..6c78eb8a9f 100644
--- a/apps/openmw/mwgui/windowmanagerimp.cpp
+++ b/apps/openmw/mwgui/windowmanagerimp.cpp
@@ -1882,7 +1882,10 @@ namespace MWGui
 
     bool WindowManager::getCursorVisible()
     {
-        return mCursorVisible && mCursorActive;
+        if (Settings::gui().mControllerMenus)
+            return mCursorVisible && mCursorActive;
+        else
+            return mCursorVisible;
     }
 
     void WindowManager::trackWindow(Layout* layout, const WindowSettingValues& settings)
