diff --git a/components/sdlutil/sdlgraphicswindow.cpp b/components/sdlutil/sdlgraphicswindow.cpp
index 9ff0545930..a84b5f90c1 100644
--- a/components/sdlutil/sdlgraphicswindow.cpp
+++ b/components/sdlutil/sdlgraphicswindow.cpp
@@ -108,17 +108,26 @@ namespace SDLUtil
         int minor = 1;
         char* ver = getenv("OPENMW_GLES_VERSION");
 
-        if (ver && strcmp(ver, "2") == 0)
+        if (ver && strcmp(ver, "20") == 0)
         {
             major = 2;
             minor = 0;
         }
-        else if (ver && strcmp(ver, "3") == 0)
+        else if (ver && strcmp(ver, "30") == 0)
+        {
+            major = 3;
+            minor = 0;
+        }
+        else if (ver && strcmp(ver, "31") == 0)
+        {
+            major = 3;
+            minor = 1;
+        }
+        else if (ver && strcmp(ver, "32") == 0)
         {
             major = 3;
             minor = 2;
         }
-
         SDL_GL_SetAttribute(SDL_GL_CONTEXT_PROFILE_MASK, SDL_GL_CONTEXT_PROFILE_ES);
         SDL_GL_SetAttribute(SDL_GL_CONTEXT_MAJOR_VERSION, major);
         SDL_GL_SetAttribute(SDL_GL_CONTEXT_MINOR_VERSION, minor);
