diff --git a/apps/openmw/mwrender/postprocessor.cpp b/apps/openmw/mwrender/postprocessor.cpp
index ff9ac2a42e..dcdc1b59e4 100644
--- a/apps/openmw/mwrender/postprocessor.cpp
+++ b/apps/openmw/mwrender/postprocessor.cpp
@@ -412,11 +410,14 @@ namespace MWRender
 
             if (mNormalsSupported)
             {
+/*
                 auto& shaderManager
                     = MWBase::Environment::get().getResourceSystem()->getSceneManager()->getShaderManager();
                 auto defines = shaderManager.getGlobalDefines();
                 defines["disableNormals"] = mNormals ? "0" : "1";
                 shaderManager.setGlobalDefines(defines);
+*/
+                mRendering.setWriteNormals(mNormals);
             }
 
             mRendering.getLightRoot()->setCollectPPLights(mPassLights);
diff --git a/apps/openmw/mwrender/renderingmanager.cpp b/apps/openmw/mwrender/renderingmanager.cpp
index 74916dcb61..861af0a2fb 100644
--- a/apps/openmw/mwrender/renderingmanager.cpp
+++ b/apps/openmw/mwrender/renderingmanager.cpp
@@ -1304,6 +1304,7 @@ namespace MWRender
                 && (it->second == "force per pixel lighting" || it->second == "classic falloff"
                     || it->second == "clamp lighting"))
             {
+/*
                 mViewer->stopThreading();
 
                 auto defines = mResourceSystem->getSceneManager()->getShaderManager().getGlobalDefines();
@@ -1311,11 +1312,13 @@ namespace MWRender
                 defines["classicFalloff"] = Settings::shaders().mClassicFalloff ? "1" : "0";
                 defines["clamp"] = Settings::shaders().mClampLighting ? "1" : "0";
                 mResourceSystem->getSceneManager()->getShaderManager().setGlobalDefines(defines);
-
+*/
                 if (MWMechanics::getPlayer().isInCell() && it->second == "classic falloff")
                     configureAmbient(*MWMechanics::getPlayer().getCell()->getCell());
 
-                mViewer->startThreading();
+//                mViewer->startThreading();
+                  mStateUpdater->reset();
+
             }
             else if (it->first == "Shaders"
                 && (it->second == "light bounds multiplier" || it->second == "maximum light distance"
@@ -1328,18 +1331,18 @@ namespace MWRender
 
                 if (it->second == "max lights")
                 {
-                    mViewer->stopThreading();
+//                    mViewer->stopThreading();
 
                     lightManager->updateMaxLights(Settings::shaders().mMaxLights);
-
+/*
                     auto defines = mResourceSystem->getSceneManager()->getShaderManager().getGlobalDefines();
                     for (const auto& [name, key] : lightManager->getLightDefines())
                         defines[name] = key;
                     mResourceSystem->getSceneManager()->getShaderManager().setGlobalDefines(defines);
-
+*/
                     mStateUpdater->reset();
 
-                    mViewer->startThreading();
+//                    mViewer->startThreading();
                 }
             }
             else if (it->first == "Post Processing" && it->second == "enabled")
@@ -1588,4 +1591,9 @@ namespace MWRender
     {
         mNavMesh->setMode(value);
     }
+
+    void RenderingManager::setWriteNormals(bool enabled)
+    {
+        mStateUpdater->setWriteNormals(enabled);
+    }
 }
diff --git a/apps/openmw/mwrender/renderingmanager.hpp b/apps/openmw/mwrender/renderingmanager.hpp
index b8ad535eb3..a6e04698b6 100644
--- a/apps/openmw/mwrender/renderingmanager.hpp
+++ b/apps/openmw/mwrender/renderingmanager.hpp
@@ -279,6 +279,8 @@ namespace MWRender
 
         void setNavMeshMode(Settings::NavMeshRenderMode value);
 
+        void setWriteNormals(bool enabled);
+
     private:
         void updateTextureFiltering();
         void updateAmbient();
diff --git a/components/shader/shadervisitor.cpp b/components/shader/shadervisitor.cpp
index a9ad457a67..2c70536861 100644
--- a/components/shader/shadervisitor.cpp
+++ b/components/shader/shadervisitor.cpp
@@ -677,7 +677,10 @@ namespace Shader
             defineMap["endLight"] = "0";
 
         if (simpleLighting || dynamic_cast<osgParticle::ParticleSystem*>(&node))
-            defineMap["forcePPL"] = "0";
+        {
+             defineMap["forcePPL"] = "0";
+            writableStateSet->setDefine("FORCE_PPL", "0", osg::StateAttribute::ON | osg::StateAttribute::PROTECTED);
+        }
 
         bool particleOcclusion = false;
         node.getUserValue("particleOcclusion", particleOcclusion);
diff --git a/components/sceneutil/stateupdater.cpp b/components/sceneutil/stateupdater.cpp
index e30fbe7a47..35eed92ef9 100644
--- a/components/sceneutil/stateupdater.cpp
+++ b/components/sceneutil/stateupdater.cpp
@@ -11,6 +11,7 @@
 #include <components/resource/scenemanager.hpp>
 #include <components/stereo/multiview.hpp>
 #include <components/stereo/stereomanager.hpp>
+#include <components/settings/values.hpp>
 
 namespace SceneUtil
 {
@@ -140,6 +141,12 @@ namespace SceneUtil
         }
         else
             stateset->removeAttribute(osg::StateAttribute::POLYGONMODE);
+
+            stateset->setDefine("WRITE_NORMALS", (mWriteNormals) ? "1" : "0", osg::StateAttribute::ON);
+            stateset->setDefine("FORCE_PPL", (Settings::shaders().mForcePerPixelLighting == true) ? "1" : "0", osg::StateAttribute::ON);
+            stateset->setDefine("CLASSIC_FALLOFF", (Settings::shaders().mClassicFalloff == true) ? "1" : "0", osg::StateAttribute::ON);
+            stateset->setDefine("CLAMP_LIGHTING", (Settings::shaders().mClampLighting == true) ? "1" : "0", osg::StateAttribute::ON);
+            stateset->setDefine("MAX_LIGHTS", std::to_string(Settings::shaders().mMaxLights), osg::StateAttribute::ON);
     }
 
     void StateUpdater::apply(osg::StateSet* stateset, osg::NodeVisitor*)
@@ -151,6 +158,8 @@ namespace SceneUtil
         fog->setColor(mFogColor);
         fog->setStart(mFogStart);
         fog->setEnd(mFogEnd);
+
+        stateset->setDefine("WRITE_NORMALS", (mWriteNormals) ? "1" : "0", osg::StateAttribute::ON);
     }
 
     void StateUpdater::setAmbientColor(const osg::Vec4f& col)
@@ -187,4 +196,9 @@ namespace SceneUtil
         return mWireframe;
     }
 
+    void StateUpdater::setWriteNormals(bool enabled)
+    {
+        mWriteNormals = enabled;
+    }
+
 }
diff --git a/components/sceneutil/stateupdater.hpp b/components/sceneutil/stateupdater.hpp
index ffdf04e140..da7235b350 100644
--- a/components/sceneutil/stateupdater.hpp
+++ b/components/sceneutil/stateupdater.hpp
@@ -96,13 +96,16 @@ namespace SceneUtil
         void setWireframe(bool wireframe);
         bool getWireframe() const;
 
+        void setWriteNormals(bool enabled);
+
     private:
         osg::Vec4f mAmbientColor;
         osg::Vec4f mFogColor;
         float mFogStart = 0.f;
         float mFogEnd = 0.f;
         bool mWireframe = false;
+        bool mWriteNormals = false;
     };
 }
 
-#endif
\ No newline at end of file
+#endif
