diff --git a/include/osg/Texture b/include/osg/Texture
index f6614633a..f17966f66 100644
--- a/include/osg/Texture
+++ b/include/osg/Texture
@@ -409,6 +409,10 @@
     #define GL_MAX_COMBINED_IMAGE_UNIFORMS    0x90CF
 #endif
 
+#define GL_CUBIC_IMG                                    0x9139
+#define GL_CUBIC_MIPMAP_NEAREST_IMG                     0x913A
+#define GL_CUBIC_MIPMAP_LINEAR_IMG                      0x913B
+
 //#define OSG_COLLECT_TEXTURE_APPLIED_STATS 1
 
 namespace osg {
@@ -506,7 +510,10 @@ class OSG_EXPORT Texture : public osg::StateAttribute
             LINEAR_MIPMAP_NEAREST     = GL_LINEAR_MIPMAP_NEAREST,
             NEAREST                   = GL_NEAREST,
             NEAREST_MIPMAP_LINEAR     = GL_NEAREST_MIPMAP_LINEAR,
-            NEAREST_MIPMAP_NEAREST    = GL_NEAREST_MIPMAP_NEAREST
+            NEAREST_MIPMAP_NEAREST    = GL_NEAREST_MIPMAP_NEAREST,
+            CUBIC                     = GL_CUBIC_IMG,
+            CUBIC_MIPMAP_LINEAR       = GL_CUBIC_MIPMAP_LINEAR_IMG,
+            CUBIC_MIPMAP_NEAREST      = GL_CUBIC_MIPMAP_NEAREST_IMG
         };
 
         /** Sets the texture filter mode. */
